File: services/python-svc/app/platform/engine/rules_engine.py
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, List, Optional
import uuid

class OutputType(str, Enum):
    SCORE = "score"
    FLAG = "flag"
    LABEL = "label"
    VALUE = "value"
    EVENT = "event"

class Operator(str, Enum):
    EQUALS = "EQUALS"
    NOT_EQUALS = "NOT_EQUALS"
    GREATER_THAN = "GREATER_THAN"
    GREATER_THAN_OR_EQUAL = "GREATER_THAN_OR_EQUAL"
    LESS_THAN = "LESS_THAN"
    LESS_THAN_OR_EQUAL = "LESS_THAN_OR_EQUAL"

class LogicalOperator(str, Enum):
    AND = "AND"
    OR = "OR"
    NOT = "NOT"

@dataclass
class AtomicRule:
    id: str
    version: int
    name: str
    owner: str
    domain: str
    rationale: str
    field: str
    operator: Operator
    value: Any
    output_key: str
    output_type: OutputType
    user_facing_explanation: str
    created_by: str
    updated_by: str
    active_from: Optional[datetime] = None
    active_to: Optional[datetime] = None
    enabled: bool = True
    priority: int = 10
    weight: float = 1.0
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    source: Optional[str] = None
    supersedes: Optional[str] = None
    change_reason: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    timeout_ms: Optional[int] = None
    deterministic: bool = True

@dataclass
class Policy:
    id: str
    name: str
    version: int
    owner: str
    domain: str
    rationale: str
    atomic_rules: List[AtomicRule]
    user_facing_explanation: str

@dataclass
class RuleEvaluationAudit:
    evaluation_id: str
    policy_id: str
    policy_version: int
    trace_id: str
    tenant_id: str
    evaluated_at: datetime
    passed: bool
    facts: Dict[str, Any]
    fired_atomic_rules: List[str]
    user_facing_reasons: List[str]
    debug_waterfall: List[str]
    rule_rationales: Dict[str, str]

class RulesEvaluator:
    def evaluate_policy(self, policy: Policy, facts: Dict[str, Any], trace_id: str, tenant_id: str) -> RuleEvaluationAudit:
        now = datetime.now(timezone.utc)
        fired_rules: List[str] = []
        reasons: List[str] = []
        waterfall: List[str] = []
        rationales: Dict[str, str] = {}
        policy_passed = True

        for rule in policy.atomic_rules:
            rationales[rule.id] = rule.rationale
            if not rule.enabled:
                waterfall.append(f"[Level 1 Atomic] Rule {rule.id} skipped (disabled)")
                continue

            fact_val = facts.get(rule.field)
            passed = str(fact_val) == str(rule.value) if fact_val is not None else False
            waterfall.append(f"[Level 1 Atomic] Rule {rule.id} ({rule.field} {rule.operator.value} {rule.value}): {passed} | Rationale: {rule.rationale}")

            if passed:
                fired_rules.append(rule.id)
            else:
                policy_passed = False
                reasons.append(rule.user_facing_explanation)

        return RuleEvaluationAudit(
            evaluation_id=f"eval-{uuid.uuid4()}",
            policy_id=policy.id,
            policy_version=policy.version,
            trace_id=trace_id,
            tenant_id=tenant_id,
            evaluated_at=now,
            passed=policy_passed,
            facts=facts,
            fired_atomic_rules=fired_rules,
            user_facing_reasons=reasons,
            debug_waterfall=waterfall,
            rule_rationales=rationales
        )

File: services/kotlin-svc/platform/engine/RulesEngine.kt
package platform.engine

import java.time.Instant

enum class OutputType { SCORE, FLAG, LABEL, VALUE, EVENT }
enum class Operator { EQUALS, NOT_EQUALS, GREATER_THAN, GREATER_THAN_OR_EQUAL, LESS_THAN, LESS_THAN_OR_EQUAL, IN, NOT_IN }
enum class LogicalOperator { AND, OR, NOT }

/** Level 1 — Atomic Rule with Full Provenance, Change Tracking & Operational Metadata */
data class AtomicRule(
    val id: String,
    val version: Int,
    val name: String,
    val owner: String,
    val domain: String,
    val rationale: String,
    val activeFrom: Instant? = null,
    val activeTo: Instant? = null,
    val enabled: Boolean = true,
    val field: String,
    val operator: Operator,
    val value: Any,
    val priority: Int = 10,
    val weight: Double = 1.0,
    val conditionName: String? = null,
    val computationName: String? = null,
    val outputKey: String,
    val outputType: OutputType,
    val userFacingExplanation: String,
    val createdAt: Instant = Instant.now(),
    val createdBy: String,
    val updatedAt: Instant = Instant.now(),
    val updatedBy: String,
    val source: String? = null,
    val supersedes: String? = null,
    val changeReason: String? = null,
    val tags: List<String> = emptyList(),
    val timeoutMs: Int? = null,
    val deterministic: Boolean = true
)

/** Level 2 — Compound Rule */
data class CompoundRule(
    val id: String,
    val version: Int,
    val name: String,
    val owner: String,
    val domain: String,
    val rationale: String,
    val enabled: Boolean = true,
    val logicalOp: LogicalOperator,
    val atomicRules: List<AtomicRule>,
    val priority: Int = 10,
    val outputKey: String,
    val outputType: OutputType,
    val weight: Double = 1.0,
    val userFacingExplanation: String
)

/** Level 3 — Policy */
data class Policy(
    val id: String,
    val name: String,
    val version: Int,
    val owner: String,
    val domain: String,
    val rationale: String,
    val atomicRules: List<AtomicRule>,
    val compoundRules: List<CompoundRule>,
    val userFacingExplanation: String
)

/** Complete Audit Trail — Accountability, Debuggability, Rationale */
data class RuleEvaluationAudit(
    val evaluationId: String,
    val policyId: String,
    val policyVersion: Int,
    val traceId: String,
    val tenantId: String,
    val evaluatedAt: Instant = Instant.now(),
    val passed: Boolean,
    val facts: Map<String, Any>,
    val firedAtomicRules: List<String>,
    val firedCompoundRules: List<String>,
    val userFacingReasons: List<String>,
    val debugWaterfall: List<String>,
    val ruleRationales: Map<String, String>
)

class RulesEvaluator {
    fun evaluatePolicy(policy: Policy, facts: Map<String, Any>, traceId: String, tenantId: String): Pair<Boolean, RuleEvaluationAudit> {
        val now = Instant.now()
        val firedAtomic = mutableListOf<String>()
        val reasons = mutableListOf<String>()
        val waterfall = mutableListOf<String>()
        val rationales = mutableMapOf<String>()
        var policyPassed = true

        policy.atomicRules.forEach { rule ->
            rationales[rule.id] = rule.rationale
            if (!rule.enabled) {
                waterfall.add("[Level 1 Atomic] Rule ${rule.id} (ver:${rule.version}) skipped (disabled)")
                return@forEach
            }
            if (rule.activeFrom != null && now.isBefore(rule.activeFrom)) {
                waterfall.add("[Level 1 Atomic] Rule ${rule.id} skipped (not active yet)")
                return@forEach
            }
            if (rule.activeTo != null && now.isAfter(rule.activeTo)) {
                waterfall.add("[Level 1 Atomic] Rule ${rule.id} skipped (expired)")
                return@forEach
            }

            val passed = evaluateAtomic(rule, facts)
            waterfall.add("[Level 1 Atomic] Rule ${rule.id} (Priority: ${rule.priority}, Weight: ${rule.weight}, ${rule.field} ${rule.operator} ${rule.value}): $passed | Rationale: ${rule.rationale}")
            if (passed) {
                firedAtomic.add(rule.id)
            } else {
                policyPassed = false
                reasons.add(rule.userFacingExplanation)
            }
        }

        val audit = RuleEvaluationAudit(
            evaluationId = "eval-\${System.currentTimeMillis()}",
            policyId = policy.id,
            policyVersion = policy.version,
            traceId = traceId,
            tenantId = tenantId,
            evaluatedAt = now,
            passed = policyPassed,
            facts = facts,
            firedAtomicRules = firedAtomic,
            firedCompoundRules = emptyList(),
            userFacingReasons = reasons,
            debugWaterfall = waterfall,
            ruleRationales = rationales
        )

        return Pair(policyPassed, audit)
    }

    private fun evaluateAtomic(rule: AtomicRule, facts: Map<String, Any>): Boolean {
        val valFact = facts[rule.field] ?: return false
        return valFact.toString() == rule.value.toString()
    }
}
